pkg/providers/aws/cluster_network_provider.go

```diff
@@ -26,77 +38,562 @@ import (
 	errorUtil "github.com/pkg/errors"
 )
 
-//todo we need to ensure we test that if `IsEnabled` is true, it remains true for consecutive reconciles
+const (
+	DefaultRHMIVpcNameTagKey   = "Name"
+	DefaultRHMIVpcNameTagValue = "RHMI Cloud Resource VPC"
+)
+
+// wrapper for ec2 vpcs, to allow for extensibility
+type Network struct {
+	Vpc *ec2.Vpc
+}
 
 type NetworkManager interface {
-	CreateNetwork() error
-	DeleteNetwork() error
-	IsEnabled(context.Context, client.Client, ec2iface.EC2API) (bool, error)
+	CreateNetwork(context.Context, *net.IPNet) (*Network, error)
+	DeleteNetwork(context.Context) error
+	IsEnabled(context.Context) (bool, error)
 }
 
 var _ NetworkManager = (*NetworkProvider)(nil)
 
 type NetworkProvider struct {
-	logger *logrus.Entry
+	Client         client.Client
+	RdsApi         rdsiface.RDSAPI
+	Ec2Api         ec2iface.EC2API
+	ElasticacheApi elasticacheiface.ElastiCacheAPI
+	Logger         *logrus.Entry
 }
 
-func NewNetworkManager(logger *logrus.Entry) *NetworkProvider {
+func NewNetworkManager(session *session.Session, client client.Client, logger *logrus.Entry) *NetworkProvider {
 	return &NetworkProvider{
-		logger: logger.WithField("provider", "standalone_network_provider"),
+		Client:         client,
+		RdsApi:         rds.New(session),
+		Ec2Api:         ec2.New(session),
+		ElasticacheApi: elasticache.New(session),
+		Logger:         logger.WithField("provider", "standalone_network_provider"),
+	}
+}
+
+/*
+CreateNetwork returns a Network type or error
+
+VPC's created by the cloud resource operator are identified by having a tag with the name `<organizationTag>/clusterID`.
+By default, `integreatly.org/clusterID`.
+
+CreateNetwork does:
+    * create a VPC with CIDR block and tag it, if a VPC does not exist,
+	* reconcile on subnets and subnet groups
+
+CreateNetwork does not:
+	* reconcile the vpc if the VPC already exist (this is to avoid potential changes to the CIDR range and unwanted/unexpected behaviour)
+*/
+func (n *NetworkProvider) CreateNetwork(ctx context.Context, vpcCidrBlock *net.IPNet) (*Network, error) {
+	logger := n.Logger.WithField("action", "CreateNetwork")
+	logger.Debug("CreateNetwork called")
+
+	clusterID, err := resources.GetClusterID(ctx, n.Client)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "error getting clusterID")
+	}
+	organizationTag := resources.GetOrganizationTag()
+
+	// check if there is cluster specific vpc already created.
+	foundVpc, err := getStandaloneVpc(n.Ec2Api, logger, clusterID, organizationTag)
+	if err != nil {
+		return nil, errorUtil.Wrap(err, "unable to get vpc")
+	}
+
+	if foundVpc != nil {
```

Comment:
So will we only setup subnets and subnet groups the first time the VPC is created?

If someone deletes a subnet or changes tags on it, we won't recreate it at the moment?

Reply:
the flow of `CreateNetwork` has been updated, if you could give it a once over :smile_cat: